### PR TITLE
Add doctests to Perceiver examples

### LINKED ISSUE #16292: Doc tests sprint
### This issue is part of our **Doc Test Sprint**. If you're interested in helping out come [join us on Discord](https://discord.gg/J8bW9u5abB) and talk with other contributors!

Docstring examples are often the first point of contact when trying out a new library! So far we haven't done a very good job at ensuring that all docstring examples work correctly in 🤗 Transformers - but we're now very dedicated to ensure that all documentation examples work correctly by integrating Python's doctest (https://docs.python.org/3/library/doctest.html) into our daily github action tests. 

In short we should add doc tests for all of our models for both PyTorch and Tensorflow. So far we've only covered ca. 30 PyTorch models as can be seen here: https://github.com/huggingface/transformers/blob/master/utils/documentation_tests.txt . 

Adding a documentation test for a model is a great way to better understand how the model works, a simple (possibly first) contribution to Transformers and most importantly a very important contribution to the Transformers community 🔥 

If you're interested in adding a documentation test, please read through the **Guide to contributing** below.

This issue is a call for contributors, to make sure docstring exmaples of existing model architectures work correctly. If you wish to contribute, reply in this thread which architectures you'd like to take :)

### Guide to contributing:
1. Ensure you've read our contributing [guidelines](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md) 📜 
2. Claim your architecture(s) in this thread (confirm no one is working on it) 🎯 
3. Implement the changes as in https://github.com/huggingface/transformers/pull/15987 (see the diff on the model architectures for a few examples) 💪 
    - The file you want to look at is in `src/transformers/models/[model_name]/modeling_[model_name].py`, `src/transformers/models/[model_name]/modeling_tf_[model_name].py` or `src/transformers/doc_utils.py` or `src/transformes/file_utils.py`
    - Make sure to run the doc example doc test locally as described in https://github.com/huggingface/transformers/tree/master/docs#for-python-files
    - Optionally, change the example docstring to a more sensible example that gives a better suited result
    - Make the test pass
    - Add the file name to https://github.com/huggingface/transformers/blob/master/utils/documentation_tests.txt (making sure the file stays in alphabetical order)
    - Run the doc example test again locally
4. Open the PR and tag me @patrickvonplaten @ydshieh or @patil-suraj (don't forget to run `make fixup` before your final commit) 🎊 
    - Note that some code is copied across our codebase. If you see a line like `# Copied from transformers.models.bert...`, this means that the code is copied from that source, and our scripts will automatically keep that in sync. If you see that, you should not edit the copied method! Instead, edit the original method it's copied from, and run make fixup to synchronize that across all the copies. Be sure you installed the development dependencies with `pip install -e ".[dev]"`, as described in the contributor guidelines above, to ensure that the code quality tools in `make fixup` can run.

### PyTorch Model Examples added to tests:
- [ ] ALBERT
- [x] BART
- [x] BEiT
- [ ] BERT
- [ ] Bert
- [ ] BigBird
- [x] BigBirdPegasus
- [x] Blenderbot
- [x] BlenderbotSmall
- [ ] CamemBERT
- [ ] Canine
- [ ] CLIP
- [ ] ConvBERT
- [x] ConvNext
- [ ] CTRL
- [x] Data2VecAudio
- [ ] Data2VecText
- [ ] DeBERTa
- [ ] DeBERTa-v2
- [x] DeiT
- [ ] DETR
- [ ] DistilBERT
- [ ] DPR
- [ ] ELECTRA
- [ ] Encoder
- [ ] FairSeq
- [ ] FlauBERT
- [ ] FNet
- [ ] Funnel
- [ ] GPT
- [ ] GPT-J
- [x] Hubert
- [ ] I-BERT
- [ ] ImageGPT
- [ ] LayoutLM
- [ ] LayoutLMv2
- [ ] LED
- [ ] Longformer
- [ ] LUKE
- [ ] LXMERT
- [ ] M2M100
- [x] Marian
- [ ] MaskFormer
- [x] mBART
- [ ] MegatronBert
- [ ] MobileBERT
- [ ] MPNet
- [ ] mT5
- [ ] Nystromformer
- [ ] OpenAI
- [ ] OpenAI
- [x] Pegasus
- [ ] Perceiver
- [x] PLBart
- [x] PoolFormer
- [ ] ProphetNet
- [ ] QDQBert
- [ ] RAG
- [ ] Realm
- [ ] Reformer
- [x] ResNet
- [ ] RemBERT
- [ ] RetriBERT
- [ ] RoBERTa
- [ ] RoFormer
- [x] SegFormer
- [x] SEW
- [x] SEW-D
- [x] SpeechEncoderDecoder
- [x] Speech2Text
- [x] Speech2Text2
- [ ] Splinter
- [ ] SqueezeBERT
- [x] Swin
- [ ] T5
- [ ] TAPAS
- [ ] Transformer-XL
- [ ] TrOCR
- [x] UniSpeech
- [x] UniSpeechSat
- [x] Van
- [x] ViLT
- [x] VisionEncoderDecoder
- [ ] VisionTextDualEncoder
- [ ] VisualBert
- [x] ViT
- [x] ViTMAE
- [x] Wav2Vec2
- [x] WavLM
- [ ] XGLM
- [ ] XLM
- [ ] XLM-RoBERTa
- [ ] XLM-RoBERTa-XL
- [ ] XLMProphetNet
- [ ] XLNet
- [ ] YOSO

### Tensorflow Model Examples added to tests:
- [ ] ALBERT
- [ ] BART
- [ ] BEiT
- [ ] BERT
- [ ] Bert
- [ ] BigBird
- [ ] BigBirdPegasus
- [ ] Blenderbot
- [ ] BlenderbotSmall
- [ ] CamemBERT
- [ ] Canine
- [ ] CLIP
- [ ] ConvBERT
- [ ] ConvNext
- [ ] CTRL
- [ ] Data2VecAudio
- [ ] Data2VecText
- [ ] DeBERTa
- [ ] DeBERTa-v2
- [ ] DeiT
- [ ] DETR
- [ ] DistilBERT
- [ ] DPR
- [ ] ELECTRA
- [ ] Encoder
- [ ] FairSeq
- [ ] FlauBERT
- [ ] FNet
- [ ] Funnel
- [ ] GPT
- [ ] GPT-J
- [ ] Hubert
- [ ] I-BERT
- [ ] ImageGPT
- [ ] LayoutLM
- [ ] LayoutLMv2
- [ ] LED
- [ ] Longformer
- [ ] LUKE
- [ ] LXMERT
- [ ] M2M100
- [ ] Marian
- [ ] MaskFormer
- [ ] mBART
- [ ] MegatronBert
- [ ] MobileBERT
- [ ] MPNet
- [ ] mT5
- [ ] Nystromformer
- [ ] OpenAI
- [ ] OpenAI
- [ ] Pegasus
- [ ] Perceiver
- [ ] PLBart
- [ ] PoolFormer
- [ ] ProphetNet
- [ ] QDQBert
- [ ] RAG
- [ ] Realm
- [ ] Reformer
- [ ] ResNet
- [ ] RemBERT
- [ ] RetriBERT
- [ ] RoBERTa
- [ ] RoFormer
- [ ] SegFormer
- [ ] SEW
- [ ] SEW-D
- [ ] SpeechEncoderDecoder
- [ ] Speech2Text
- [ ] Speech2Text2
- [ ] Splinter
- [ ] SqueezeBERT
- [ ] Swin
- [ ] T5
- [ ] TAPAS
- [ ] Transformer-XL
- [ ] TrOCR
- [ ] UniSpeech
- [ ] UniSpeechSat
- [ ] Van
- [ ] ViLT
- [ ] VisionEncoderDecoder
- [ ] VisionTextDualEncoder
- [ ] VisualBert
- [ ] ViT
- [ ] ViTMAE
- [ ] Wav2Vec2
- [ ] WavLM
- [ ] XGLM
- [ ] XLM
- [ ] XLM-RoBERTa
- [ ] XLM-RoBERTa-XL
- [ ] XLMProphetNet
- [ ] XLNet
- [ ] YOSO

### PR DESCRIPTION
**Pull request by stevenmanton**
# What does this PR do?

Related to #16292

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Taken from #16292 : @patrickvonplaten @ydshieh @patil-suraj

### CODE REVIEW

**Comment by ydshieh** [comment]
Hi @stevenmanton 

Thank you a lot for the PR! Currently the test will fail for the following 
```bash
FAILED src/transformers/models/perceiver/modeling_perceiver.py::transformers.models.perceiver.modeling_perceiver.PerceiverForImageClassificationConvProcessing.forward
FAILED src/transformers/models/perceiver/modeling_perceiver.py::transformers.models.perceiver.modeling_perceiver.PerceiverForImageClassificationFourier.forward
FAILED src/transformers/models/perceiver/modeling_perceiver.py::transformers.models.perceiver.modeling_perceiver.PerceiverForImageClassificationLearned.forward
```
We should add the expected outputs in the doc examples.
For other doc examples in this model, the tests pass, but they don't have outputs to test.
For example, `PerceiverForOpticalFlow` has
```
>>> logits = outputs.logits
```
without output and the expected value. We can have something like

```python
>>> list(logits .shape)
expected shapes
```
Would you like to fix and enhance those doc examples?

Once you have a change (staged or commited), you can run the test like
```
python utils/prepare_for_doc_test.py src/transformers/models/perceiver/modeling_perceiver.py
pytest --doctest-modules src/transformers/models/perceiver/modeling_perceiver.py -sv --doctest-continue-on-failure
```
Once the test is run, you can clean up the git status before further changes or push.

Thanks!

**Comment by stevenmanton** [comment]
@ydshieh thanks for the quick feedback! Yes, I noticed those tests were failing, but I thought it might be something about my local environment and the extra newline stuff. I just pushed a fix, which passes locally.

By the way, how did you know those tests were failing? The CI pipelines all seemed to be passing. Did you have to checkout my branch and run it locally?

**Comment by ydshieh** [comment]
@stevenmanton Thanks for the push. However, instead of changing to

```python
>>> predicted_class = model.config.id2label[predicted_class_idx]
```
we should change it to 
```python
>>> print("Predicted class:", model.config.id2label[predicted_class_idx])
add some output here - so the doctest will test again it
```
You can find similar work is done 
https://github.com/huggingface/transformers/blob/d5848a574a3990c95f20512673ecef9f57e0fe81/src/transformers/models/deit/modeling_deit.py#L735-L736

Otherwise, the example has nothing to be tested.

> By the way, how did you know those tests were failing? The CI pipelines all seemed to be passing. Did you have to checkout my branch and run it locally?

Yes. The PR CI on CircleCI does not run doctest :-). It is run after the PR being merged.

**Comment by stevenmanton** [comment]
@ydshieh Thanks for your feedback and patience. I believe I've corrected it. The extra newline stuff is confusing, but if you run `prepare_for_doc_test.py` (which I think just adds a newline to all the docstrings) on the last commit, all tests pass for me locally.

**Comment by ydshieh** [comment]
Thanks! As we are going to run the doctest for this model, would you like to add some expected outputs at the following places? 

https://github.com/huggingface/transformers/blob/bebcb950c7ad4dc1ef676806a6ac4283df7f5885/src/transformers/models/perceiver/modeling_perceiver.py#L1921

https://github.com/huggingface/transformers/blob/bebcb950c7ad4dc1ef676806a6ac4283df7f5885/src/transformers/models/perceiver/modeling_perceiver.py#L1695

https://github.com/huggingface/transformers/blob/bebcb950c7ad4dc1ef676806a6ac4283df7f5885/src/transformers/models/perceiver/modeling_perceiver.py#L1131

https://github.com/huggingface/transformers/blob/bebcb950c7ad4dc1ef676806a6ac4283df7f5885/src/transformers/models/perceiver/modeling_perceiver.py#L1033

https://github.com/huggingface/transformers/blob/bebcb950c7ad4dc1ef676806a6ac4283df7f5885/src/transformers/models/perceiver/modeling_perceiver.py#L1021

And a few places in 

https://github.com/huggingface/transformers/blob/bebcb950c7ad4dc1ef676806a6ac4283df7f5885/src/transformers/models/perceiver/modeling_perceiver.py#L1695

For `logits`, it would look like to add (the values should be collected from the run)
```
        >>> list(logits.shape)
        [1, 196, 8192]
```

**Comment by ydshieh** [comment]
When running `prepare_for_doc_test.py`, it will add some empty lines - to make doctest pass. That is why we should stage our change, run that script, run doctest, and discard the change before commit or further changes :-)

**Comment by stevenmanton** [comment]
@ydshieh Ok, I added some more checks for the sizes of logits. They all pass for me locally.

**Review by ydshieh — APPROVED** [review]
Hi @stevenmanton Thanks a lot of this PR and the follow-up effort :-). It's great, and I just added an extra expected loss value for `PerceiverForMaskedLM`.

@sgugger As I am not sure if this PR is classified as a PR that I can merge myself, tagging you for a review or let me know if I can merge the doctest PR in the future. Thanks.

**Review by sgugger — APPROVED** [review]
Thanks a lot for this PR!

@ydshieh Always good to have another pair of eyes :-)
Don't hesitate to merge such PRs without my approval in the future if I take too long to review, but I'd appreciate the opportunity to review!
